src/heap.rs

```diff
@@ -73,44 +70,63 @@ impl Heaps {
     pub(crate) fn new(calldata: Vec<u8>) -> Self {
         // The first heap can never be used because heap zero
         // means the current heap in precompile calls
-        Self(vec![
-            Heap(vec![]),
-            Heap(calldata),
-            Heap(vec![]),
-            Heap(vec![]),
-        ])
+        Self {
+            heaps: vec![Heap(vec![]), Heap(calldata), Heap(vec![]), Heap(vec![])],
+            bootloader_heap_rollback_info: vec![],
+        }
     }
 
     pub(crate) fn allocate(&mut self) -> HeapId {
-        let id = HeapId(self.0.len() as u32);
-        self.0
-            .push(Heap(vec![0; NEW_FRAME_MEMORY_STIPEND as usize]));
+        self.allocate_inner(vec![0; NEW_FRAME_MEMORY_STIPEND as usize])
+    }
+
+    pub(crate) fn allocate_with_content(&mut self, content: &[u8]) -> HeapId {
+        self.allocate_inner(content.to_vec())
+    }
+
+    fn allocate_inner(&mut self, memory: Vec<u8>) -> HeapId {
+        let id = HeapId(self.heaps.len() as u32);
+        self.heaps.push(Heap(memory));
         id
     }
 
     pub(crate) fn deallocate(&mut self, heap: HeapId) {
-        self.0[heap.0 as usize].0 = vec![];
+        self.heaps[heap.0 as usize].0 = vec![];
+    }
+
+    pub fn write_u256(&mut self, heap: HeapId, start_address: u32, value: U256) {
+        if heap == FIRST_HEAP {
```

Comment:
Just to understand better: Is it always the case that *all* writes to the bootloader heap happen when there's a living snapshot (i.e., need to be journaled)?

Reply:
That's a good point. This is the case only in sequencer I think, though sequencer is of course the top priority for performance.

Reply:
At least history can be cleared when the last snapshot is forgotten. I think that happens when a TX succeeds.